java/ql/lib/semmle/code/java/frameworks/MyBatis.qll

```diff
@@ -102,3 +102,40 @@ class MyBatisSqlOperationAnnotationMethod extends Method {
 class TypeParam extends Interface {
   TypeParam() { this.hasQualifiedName("org.apache.ibatis.annotations", "Param") }
 }
+
+module ProviderInjection {
+  private import semmle.code.java.dataflow.DataFlow
+
+  class MyBatisInjectionSink extends DataFlow::Node {
+    MyBatisInjectionSink() {
+      exists(Annotation a, Method m, TypeLiteral type, Class c |
+        a.getType()
+            .hasQualifiedName("org.apache.ibatis.annotations",
+              ["Select", "Delete", "Insert", "Update"] + "Provider") and
+        type = a.getValue(["type", "value"]) and
+        c.hasMethod(m, type.getTypeName().getType()) and
+        m.hasName(a.getTarget().getName()) and
+        this.asExpr() = m.getBody().getAStmt().(ReturnStmt).getResult()
+      )
+    }
+  }
+
+  class MyBatisAbstractSQLStep extends Unit {
+    predicate step(DataFlow::Node node1, DataFlow::Node node2) {
+      exists(MethodAccess ma |
+        ma.getMethod()
+            .getDeclaringType()
+            .hasQualifiedName("org.apache.ibatis.jdbc", ["AbstractSQL", "AbstractSQL<SQL>"]) and
```

Comment:
In case you decide to go the `AdditionalTaintStep` route: instead of having to write the type parameter here, you can just use `getSourceDeclaration`:

```suggestion
            .getDeclaringType()
            .getSourceDeclaration()
            .hasQualifiedName("org.apache.ibatis.jdbc", "AbstractSQL") and
```